### PR TITLE
DeletedMessageVault Export Integration Tests

### PR DESCRIPTION
**Pull request by trantienduchn**

### CODE REVIEW

File: mailbox/backup/src/test/java/org/apache/james/mailbox/backup/ZipAssert.java [hunk 1]
```diff
@@ -94,6 +97,10 @@ private static BasicErrorMessageFactory shouldBeEmpty(ZipFile zipFile) {
         return new BasicErrorMessageFactory("%nExpecting %s to be empty", zipFile);
     }
 
+    private static BasicErrorMessageFactory shouldEqual(ZipFile zipFile, ZipFile expectedZipFile) {
```
Comment (chibenwa) [inline mailbox/backup/src/test/java/org/apache/james/mailbox/backup/ZipAssert.java]: `shouldBeEqual`

File: mailbox/backup/src/test/java/org/apache/james/mailbox/backup/ZipAssert.java [hunk 2]
```diff
@@ -127,4 +134,24 @@ public ZipAssert hasNoEntry() {
...
 
+    public ZipAssert hasSameContentWith(ZipFile anotherZipFile) throws IOException {
+        if (zipFile == null || anotherZipFile == null) {
+            throwAssertionError(shouldEqual(zipFile, anotherZipFile));
+        }
+        ArrayList<ZipArchiveEntry> entries = Collections.list(zipFile.getEntries());
+        ArrayList<ZipArchiveEntry> entriesOfAnother = Collections.list(anotherZipFile.getEntries());
+
+        if (entries.size() != entriesOfAnother.size()) {
+            throwAssertionError(shouldEqual(zipFile, anotherZipFile));
```
Comment (chibenwa) [inline mailbox/backup/src/test/java/org/apache/james/mailbox/backup/ZipAssert.java]: You should have a nice message telling that the two zip don't have the same number of entries

File: mailbox/backup/src/test/java/org/apache/james/mailbox/backup/ZipAssert.java [hunk 3]
```diff
@@ -127,4 +134,24 @@ public ZipAssert hasNoEntry() {
...
+        }
+
+        for (int i = 0; i < entries.size(); i++) {
+            ZipArchiveEntry entry = entries.get(i);
+            ZipArchiveEntry entryOfAnother = entriesOfAnother.get(i);
+            assertThat(zipFile.getInputStream(entry))
+                .hasSameContentAs(zipFile.getInputStream(entryOfAnother));
```
Comment (chibenwa) [inline mailbox/backup/src/test/java/org/apache/james/mailbox/backup/ZipAssert.java]: We should have a description that `entry n°36 is not the same" `

We should also display entries name

File: mailbox/backup/src/test/java/org/apache/james/mailbox/backup/ZipAssert.java [hunk 4]
```diff
@@ -20,7 +20,10 @@
 package org.apache.james.mailbox.backup;
```
Comment (chibenwa) [inline mailbox/backup/src/test/java/org/apache/james/mailbox/backup/ZipAssert.java]: ZipAssert should be in a different commit IMO

File: mailbox/backup/src/test/java/org/apache/james/mailbox/backup/ZipAssert.java [hunk 5]
```diff
@@ -127,4 +134,24 @@ public ZipAssert hasNoEntry() {
         return myself;
     }
 
+    public ZipAssert hasSameContentWith(ZipFile anotherZipFile) throws IOException {
```
Comment (chibenwa) [inline mailbox/backup/src/test/java/org/apache/james/mailbox/backup/ZipAssert.java]: No tests ?

File: server/protocols/jmap-integration-testing/cassandra-jmap-integration-testing/src/test/java/org/apache/james/jmap/cassandra/CassandraDeletedMessageVaultTest.java [hunk 6]
```diff
@@ -49,7 +52,8 @@ protected GuiceJamesServer createJmapServer() throws IOException {
                     PreDeletionHookConfiguration.forClass(DeletedMessageVaultHook.class))),
             binder -> binder.bind(WebAdminConfiguration.class).toInstance(WebAdminConfiguration.TEST_CONFIGURATION),
             binder -> binder.bind(MailRepositoryDeletedMessageVault.Configuration.class)
-                .toInstance(new MailRepositoryDeletedMessageVault.Configuration(MailRepositoryUrl.from("cassandra://var/deletedMessages/user"))));
+                .toInstance(new MailRepositoryDeletedMessageVault.Configuration(MailRepositoryUrl.from("cassandra://var/deletedMessages/user"))),
+            binder -> Multibinder.newSetBinder(binder, GuiceProbe.class).addBinding().to(FileSystemProbeImpl.class));
```
Comment (chibenwa) [inline server/protocols/jmap-integration-testing/cassandra-jmap-integration-testing/src/test/java/org/apache/james/jmap/cassandra/CassandraDeletedMessageVaultTest.java]: Why is the FileSystemProbeImpl needed here?

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/DeletedMessagesVaultTest.java [hunk 7]
```diff
@@ -68,6 +78,48 @@
 import io.restassured.specification.RequestSpecification;
 
 public abstract class DeletedMessagesVaultTest {
+
+    public static class FileSystemProbeImpl implements GuiceProbe {
```
Comment (chibenwa) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/DeletedMessagesVaultTest.java]: IMO wrong way to do it.

We should rather override the filesystem component with our own.

I propose that the `createJmapServer` takes a FileSystem argument

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/DeletedMessagesVaultTest.java [hunk 8]
```diff
@@ -410,6 +465,232 @@ public void vaultEndpointShouldRestoreMessageForSharer() {
...
+
+        exportVaultContent(ExportRequest
+            .userExportFrom(HOMER)
+            .sharee(BART)
+            .query(MATCH_ALL_QUERY));
+
+        WAIT_TWO_MINUTES.until(() -> listMessageIdsForAccount(bartAccessToken).size() == 2);
+        String exportingMessageId = listMessageIdsForAccount(bartAccessToken).get(0);
+        String fileLocation =
+            with()
```
Comment (chibenwa) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/DeletedMessagesVaultTest.java]: Extract this complex JMAP query behind an explicit method: `bartReceivesTheSharingTheExportMail`

**Comment by trantienduchn** [comment]
```
[1d4e7d9c3a2e0cdead01e01cfb65f3cea46175e0] [ERROR]   DeletedMessagesVaultTest.tearDown:178 » FileNotFound File does not exist: ../D...
[1d4e7d9c3a2e0cdead01e01cfb65f3cea46175e0] [ERROR]   DeletedMessagesVaultTest.tearDown:178 » FileNotFound File does not exist: ../D...
[1d4e7d9c3a2e0cdead01e01cfb65f3cea46175e0] [ERROR]   DeletedMessagesVaultTest.tearDown:178 » FileNotFound File does not exist: ../D...
[1d4e7d9c3a2e0cdead01e01cfb65f3cea46175e0] [ERROR]   DeletedMessagesVaultTest.vaultExportShouldSendMailToShareeWhenImapDeletedMailbox:524->zipFile:759 » Zip
[1d4e7d9c3a2e0cdead01e01cfb65f3cea46175e0] [ERROR]   DeletedMessagesVaultTest.vaultExportShouldSendMailsToShareeWhenImapDeleteMessage:499->zipFile:759 » Zip
[1d4e7d9c3a2e0cdead01e01cfb65f3cea46175e0] [ERROR]   DeletedMessagesVaultTest.vaultExportShouldSendMailsToShareeWhenJmapDeleteMessage:476->zipFile:759 » Zip
```

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/DeletedMessagesVaultTest.java [hunk 9]
```diff
@@ -114,12 +165,17 @@ public void setup() throws Throwable {
         homerAccessToken = authenticateJamesUser(baseUri(jmapServer), HOMER, PASSWORD);
         bartAccessToken = authenticateJamesUser(baseUri(jmapServer), BART, BOB_PASSWORD);
 
-        webAdminApi = WebAdminUtils.spec(jmapServer.getProbe(WebAdminGuiceProbe.class).getWebAdminPort());
+        WebAdminGuiceProbe webAdminGuiceProbe = jmapServer.getProbe(WebAdminGuiceProbe.class);
+        webAdminApi = given()
+            .spec(WebAdminUtils
+                .buildRequestSpecification(webAdminGuiceProbe.getWebAdminPort())
+                .build());
```
Comment (chibenwa) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/DeletedMessagesVaultTest.java]: `webAdminApi = WebAdminUtils.spec(jmapServer.getProbe(WebAdminGuiceProbe.class).getWebAdminPort());`

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/DeletedMessagesVaultTest.java [hunk 10]
```diff
@@ -405,6 +461,172 @@ public void vaultEndpointShouldRestoreMessageForSharer() {
...
+        bartSendMessageToHomer();
+        WAIT_TWO_MINUTES.until(() -> listMessageIdsForAccount(homerAccessToken).size() == 1);
+        String messageIdOfHomer = listMessageIdsForAccount(homerAccessToken).get(0);
+
+        homerDeletesMessages(listMessageIdsForAccount(homerAccessToken));
+        WAIT_TWO_MINUTES.until(() -> listMessageIdsForAccount(homerAccessToken).size() == 0);
+
+        String fileLocation = exportAndGetFileLocationFromLastMail(EXPORT_ALL_HOMER_MESSAGES_TO_BART, bartAccessToken);
+
+        try (ZipFile zipFile = zipFile(fileSystem.getResource(fileLocation))) {
```
Comment (chibenwa) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/DeletedMessagesVaultTest.java]:
```
try (ZipAssert zipAssert = assertThatZip(fileSystem.getResource(fileLocation)) {
    zipAssert.containsOnlyEntriesMatching(hasName(messageIdOfHomer));
}
```

We may need to write an `assertThatZip` method taking an inputstream and doing an inmemory copy...

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/DeletedMessagesVaultTest.java [hunk 11]
```diff
@@ -518,4 +740,22 @@ private void homerMovesTheMailInAnotherMailbox(String messageId) {
...
+            .jsonPath()
+                .get("taskId");
+
+        webAdminApi.given()
+                .get("/tasks/" + taskId + "/await")
+            .then()
+                .body("status", is("completed"));
+    }
+
+    private ZipFile zipFile(InputStream data) throws IOException {
```
Comment (chibenwa) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/DeletedMessagesVaultTest.java]: This technical method should be abstracted away by ZipAssert related abstractions

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/DeletedMessagesVaultTest.java [hunk 12]
```diff
@@ -405,6 +461,172 @@ public void vaultEndpointShouldRestoreMessageForSharer() {
...
+        homerDeletesMessages(listMessageIdsForAccount(homerAccessToken));
+        WAIT_TWO_MINUTES.until(() -> listMessageIdsForAccount(homerAccessToken).size() == 0);
+
+        ExportRequest exportRequest = ExportRequest
+            .userExportFrom(HOMER)
+            .exportTo(BART)
+            .query("{" +
+                "  \"fieldName\": \"subject\"," +
+                "  \"operator\": \"equals\"," +
+                "  \"value\": \"third subject\"" +
```
Comment (chibenwa) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/DeletedMessagesVaultTest.java]: I would name this `"value": "non-matching"` to make it very clear

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/DeletedMessagesVaultTest.java [hunk 13]
```diff
@@ -114,12 +165,17 @@ public void setup() throws Throwable {
...
+            .spec(WebAdminUtils
+                .buildRequestSpecification(webAdminGuiceProbe.getWebAdminPort())
+                .build());
     }
 
     @After
-    public void tearDown() {
+    public void tearDown() throws Exception {
         jmapServer.stop();
+        FileUtils.forceDelete(fileSystem.getBasedir());
```
Comment (chibenwa) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/DeletedMessagesVaultTest.java]: Can we : 
 - Move `org.apache.james.blob.export.file.FileSystemExtension` into server-core test jar?
 - Use this extention here

Comment (trantienduchn) [reply server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/DeletedMessagesVaultTest.java]: I can move it, but currently, DeletedMessagesVaultTest is using Junit 4, It will take effort to migrate this class to Junit 5

Comment (chibenwa) [reply server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/DeletedMessagesVaultTest.java]: Crap, good point. That's OK for now

File: mailbox/backup/src/test/java/org/apache/james/mailbox/backup/ZipAssert.java [hunk 14]
```diff
@@ -93,6 +99,29 @@ private static BasicErrorMessageFactory shouldBeEmpty(ZipFile zipFile) {
         return new BasicErrorMessageFactory("%nExpecting %s to be empty", zipFile);
     }
 
+    private static BasicErrorMessageFactory shouldBeNonNull() {
+        return new BasicErrorMessageFactory("%nExpecting zipFile to be non-null");
+    }
+
+    private static BasicErrorMessageFactory shouldHaveSameEntriesSize(List<ZipArchiveEntry> entries, List<ZipArchiveEntry> expectedEntries) {
+        return new BasicErrorMessageFactory("%nExpecting zipFile entries has size (%s) but actually (%s)", expectedEntries.size(), entries.size());
```
Comment (mbaechler) [inline mailbox/backup/src/test/java/org/apache/james/mailbox/backup/ZipAssert.java]: Expecting zipFile to contain %s entries (%s) but actually contains (%s)

File: mailbox/backup/src/test/java/org/apache/james/mailbox/backup/ZipAssert.java [hunk 15]
```diff
@@ -93,6 +99,29 @@ private static BasicErrorMessageFactory shouldBeEmpty(ZipFile zipFile) {
...
+    }
+
+    private static ErrorMessageFactory entriesShouldHaveSameContentAt(int entryIndex, String entryContentAssertionError) {
+        return new BasicErrorMessageFactory("%nExpecting zipFile entry at index %s have same content. Details: \n%s",
```
Comment (mbaechler) [inline mailbox/backup/src/test/java/org/apache/james/mailbox/backup/ZipAssert.java]: Expecting zipFile entry at index %s has same content

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/DeletedMessagesVaultTest.java (hunk 13 above)
Comment (mbaechler) [reply server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/DeletedMessagesVaultTest.java]: you can use TemporaryFolder from junit 4 rules

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/DeletedMessagesVaultTest.java [hunk 16]
```diff
@@ -405,6 +461,172 @@ public void vaultEndpointShouldRestoreMessageForSharer() {
             .body(ARGUMENTS + ".list.subject", hasItem(SUBJECT));
     }
 
+    @Category(BasicFeature.class)
+    @Test
+    public void vaultExportShouldSendMailsToShareeWhenJmapDeleteMessage() throws Exception {
```
Comment (mbaechler) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/DeletedMessagesVaultTest.java]: you don't check that it sends a mail. You check that an zip is produced.

**Comment by Arsnael** [comment]
Rebase please

**Comment by rouazana** [comment]
merged
